Source Management API

vadimjustus opened the pull request:
<!--- Provide a general summary of the Pull Request in the Title above -->

### Description
Changes which were created during the 1st contribution day in Leipzig

### Fixed Issues (if relevant)
1. magento-engcom/magento2#24: Implementation differes to [conception](https://github.com/magento-engcom/magento2/wiki/Source-WebAPI): Permissions PUT and POST permissions are both set to `source_edit` (approved by @maghamed)
2. magento-engcom/magento2#25: Stub modules Magento_InventoryApi and Magento_Inventory. Implement interfaces, data models, resource models and repository.
3. magento-engcom/magento2#7: Defining of Source API

### Manual testing scenarios
Currently code review only

### Contribution checklist
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds on Travis CI are green)

Commit:
Add new InventoryApi Module and implement Interfaces

vadimjustus committed:
magento-engcom/magento2#25 Update dependencies

Commit:
magento-engcom/magento2#25 Add new Inventory Module

vadimjustus committed:
magento-engcom/magento2#25: Update dependencies between inventory and inventory-api module

vadimjustus committed:
magento-engcom/magento2#25 remove chagnes from .gitignore file

vadimjustus committed:
magento-engcom/magento2#24: Implement web API route configuration

ishakhsuvarov committed:
magento-engcom/magento2#25: Source API Implementation

vadimjustus committed:
magento-engcom/magento2#24: Move ACL and WebApi routes to InterfaceApi module

vadimjustus committed:
magento-engcom/magento2#25: Remove \Magento\InventoryApi\Api\Data\SourceExtensionInterface as it will be autogenerated

vadimjustus committed:
Merge branch 'mni/source-management' of github.com:magento-engcom/magento2 into mni/source-management

ishakhsuvarov committed:
magento-engcom/magento2#25: Source API Implementation

nhp committed:
magento-engcom/magento2#7 Add missing empty interface class

nhp committed:
Filled SourceCarrierLinkInterface with some methods

Commit:
magento-engcom/magento2#25 Implemented Resourcemodel, Collection and part of the businessmodel for source

Commit:
Merge branch 'mni/source-management' of https://github.com/magento-engcom/magento2 into mni/source-management

vadimjustus committed:
magento-engcom/magento2#25: Implement repository model for inventory sources

vadimjustus committed:
Merge branch 'mni/source-management' of github.com:magento-engcom/magento2 into mni/source-management

vadimjustus committed:
magento-engcom/magento2#25: Implement get() method in SourceRepository

vadimjustus committed:
magento-engcom/magento2#25: Change protected properties to private

Commit:
magento-engcom/magento2#25: Source API Implementation

 - fixed code style issues

vadimjustus committed:
magento-engcom/magento2#25: Update docblock for method in order to follow the inherit doc from interface

vadimjustus committed:
magento-engcom/magento2#25: Update interfaces in order to use USE statements

vadimjustus committed:
Merge branch 'mni/source-management' of github.com:magento-engcom/magento2 into mni/source-management

Commit:
Wip

Commit:
magento-engcom/magento2#25 Implemented Resourcemodel, Collection and part of the businessmodel for sourcecarrierlink, updated carrierlink-interface, implemented part of source-businessmodel

Commit:
magento-engcom/magento2#25 solved mergeconfilcts

okorshenko commented on app/code/Magento/Inventory/Model/Source.php:
```diff
@@ -322,15 +322,16 @@ public function setPriority($priority)
     public function getCarrierLinks()
     {
-        // TODO: Implement getCarrierLinks() method.
+        return $this->getData(SourceInterface::PRIORITY);
```
Should it be a SourceInterface::CARRIER_LINKS ?

Commit:
Merge branch 'mni/source-management' of https://github.com/magento-engcom/magento2 into mni/source-management

Commit:
Merge branch 'mni/source-management' of https://github.com/magento-engcom/magento2 into mni/source-management

ishakhsuvarov committed:
magento-engcom/magento2#25: Source API Implementation

 - Created stub for api-functional test for SourceRepository

vadimjustus committed:
magento-engcom/magento2#25: Implement SourceSearchResults

vadimjustus committed:
magento-engcom/magento2#25: Add di.xml and define preferences for interfaces

vadimjustus committed:
magento-engcom/magento2#25: Implement unit tests for SourceRepository

vadimjustus committed:
Merge branch 'mni/source-management' of github.com:magento-engcom/magento2 into mni/source-management

Commit:
magento-engcom/magento2#25 added unit tests for source and sourcecarrierlink models

Commit:
Merge branch 'mni/source-management' of https://github.com/magento-engcom/magento2 into mni/source-management

ishakhsuvarov committed:
magento-engcom/magento2#25: Source API Implementation

 - Created stub for api-functional test for SourceRepository

Commit:
magento-engcom/magento2#25 added fixed typo in test and added di configuration

Commit:
magento-engcom/magento2#25 fixed unit tests for source and sourcecarrierlink

Commit:
Merge remote-tracking branch 'origin/mni/source-management-api-tests' into mni/source-management

vadimjustus committed:
magento-engcom/magento2#25: Implement rests for SourceRepository

vadimjustus committed:
Merge branch 'mni/source-management' of github.com:magento-engcom/magento2 into mni/source-management

vadimjustus committed:
magento-engcom/magento2#25: Fix unit tests

vadimjustus committed:
magento-engcom/magento2#25: Code styles

maghamed commented on app/code/Magento/Inventory/Model/Resource/Source/Collection.php:
```diff
@@ -0,0 +1,34 @@
+use Magento\Inventory\Model\Source as SourceModel;
+
+/**
+ * Class Collection
```
please provide more meaningful description

maghamed commented on app/code/Magento/Inventory/Model/Resource/Source/Collection.php:
```diff
@@ -0,0 +1,34 @@
+
+/**
+ * Class Collection
+ * @package Magento\Inventory\Model\Resource\Source
```
what the sense to use @ package here?

maghamed commented on app/code/Magento/Inventory/Model/Resource/Source/Collection.php:
```diff
@@ -0,0 +1,34 @@
+    /**
+     * @inheritdoc
+     */
+    public function getIdFieldName()
```
why do we need this method in the scope of collection? 
we usually have one for Resource Models

Tominator87 replied:
fixed

maghamed commented on app/code/Magento/Inventory/Model/Resource/SourceCarrierLink.php:
```diff
@@ -0,0 +1,27 @@
+     *
+     * @return void
+     *
+     * @codingStandardsIgnore
```
why do we have coding standards ignore for this method?

maghamed commented on app/code/Magento/Inventory/Model/Resource/SourceCarrierLink/Collection.php:
```diff
@@ -0,0 +1,36 @@
+     * @return void
+     *
+     * @codingStandardsIgnore
+     * @codeCoverageIgnore
```
why do we have coding standards ignored?

maghamed commented on app/code/Magento/Inventory/Model/Resource/SourceCarrierLink/Collection.php:
```diff
@@ -0,0 +1,36 @@
+use Magento\Framework\Model\ResourceModel\Db\Collection\AbstractCollection;
+use Magento\Inventory\Model\SourceCarrierLink as SourceCarrierLinkModel;
+
+class Collection extends AbstractCollection
```
why do we need SourceCarrierLink\Collection? 
I would prefer not to add this class now, while we don't have specific use-cases for it

maghamed commented on app/code/Magento/Inventory/Model/Source.php:
```diff
@@ -0,0 +1,353 @@
+use \Magento\InventoryApi\Api\Data\SourceInterface;
+
+/**
+ * Class Source
```
please  provide more meaningful description

maghamed commented on app/code/Magento/Inventory/Model/Source.php:
```diff
@@ -0,0 +1,353 @@
+
+/**
+ * Class Source
+ * @package Magento\Inventory\Model
```
don't use  @ package doc block

Tominator87 replied:
fixed

maghamed commented on app/code/Magento/Inventory/Model/Source.php:
```diff
@@ -0,0 +1,353 @@
+    /**
+     * @inheritdoc
+     */
+    protected $_collectionName = \Magento\Inventory\Model\Resource\Source\Collection::class;
```
it's better to use _init method instead of initializing this variable

Tominator87 replied:
thanks for the advice, fixed

maghamed commented on app/code/Magento/Inventory/Model/SourceCarrierLink.php:
```diff
@@ -0,0 +1,85 @@
+     * @codingStandardsIgnore
+     * @var string
+     */
+    protected $_collectionName = 'Magento\Inventory\Model\Resource\SourceCarrierLink\Collection';
```
it's better to use _init method in pseudo constructor instead of initializing this variable 

```php
        $this->_resourceName = $resourceName;
        if ($collectionName === null) {
            $collectionName = $resourceName . '\\' . 'Collection';
        }
        $this->_collectionName = $collectionName;
```
thus we have a convention to name a Collection the same way as Resource model with adding suffix Collection

maghamed commented on app/code/Magento/Inventory/Model/SourceRepository.php:
```diff
@@ -0,0 +1,127 @@
+
+/**
+ * Class SourceRepository
+ * @package Magento\Inventory\Model
```
no need to use @ package

Tominator87 replied:
fixed

maghamed commented on app/code/Magento/Inventory/Model/SourceRepository.php:
```diff
@@ -0,0 +1,127 @@
+     */
+    public function __construct(
+        ResourceSource $resource,
+        SourceFactory $sourceFactory,
```
it's better to use SourceFactoryInterface

Tominator87 replied:
fixed

maghamed commented on app/code/Magento/Inventory/Model/SourceRepository.php:
```diff
@@ -0,0 +1,127 @@
+        try {
+            $this->resource->save($source);
+        } catch (\Exception $exception) {
+            throw new CouldNotSaveException(__($exception->getMessage()));
```
it's an incorrect way of proxying exception message from low level exception to hight level.
Because in this case:
1. you just show all error messages happened to use. Which is security vulnerability 
2. there is no guarantee that message would be translated

Tominator87 replied:
changed it, look if this fits your expectations

maghamed commented on app/code/Magento/Inventory/Model/SourceRepository.php:
```diff
@@ -0,0 +1,127 @@
+        $this->resource->load($model, SourceInterface::SOURCE_ID, $sourceId);
+
+        if (!$model->getSourceId()) {
+            throw NoSuchEntityException::singleField(SourceInterface::SOURCE_ID, $sourceId);
```
throw new

Tominator87 replied:
` public static function singleField($fieldName, $fieldValue)
    {
        return new self(
            new Phrase(
                'No such entity with %fieldName = %fieldValue',
                [
                    'fieldName' => $fieldName,
                    'fieldValue' => $fieldValue
                ]
            )
        );
    }`
Can you look at this one? I think the static call makes sense, but i can change it if you want.

maghamed commented on app/code/Magento/Inventory/Setup/InstallSchema.php:
```diff
@@ -0,0 +1,228 @@
+
+    /**
+     *
+     */
```
please provide description

Tominator87 replied:
added description. Look if it fits your expectations

maghamed commented on app/code/Magento/Inventory/Setup/InstallSchema.php:
```diff
@@ -0,0 +1,228 @@
+
+/**
+ * Class InstallSchema
+ * @package Magento\Inventory\Setup
```
package no needed

Tominator87 replied:
fixed

Commit:
magento-engcom/magento2#25: Source API Implementation

 - code style fixes

maghamed commented on app/code/Magento/Inventory/Model/Source.php:
```diff
@@ -0,0 +1,353 @@
+use \Magento\InventoryApi\Api\Data\SourceInterface;
+
+/**
+ * Class Source
```
please provide meaningful description for class

Tominator87 replied:
added description. Look if it fits your expectations

maghamed commented on app/code/Magento/Inventory/Model/SourceCarrierLink.php:
```diff
@@ -0,0 +1,83 @@
+     * @codingStandardsIgnore
+     * @var string
+     */
+    protected $_collectionName = \Magento\Inventory\Model\Resource\SourceCarrierLink\Collection::class;
```
why do we need to have SourceCarrierLink collection?

Tominator87 replied:
fixed it

maghamed commented on app/code/Magento/Inventory/Model/SourceRepository.php:
```diff
@@ -0,0 +1,120 @@
+    private $sourceSearchResultsFactory;
+
+    /**
+     * SourceRepository constructor.
```
we usually make a one line break between 
class description and parameter list

Also please provide meaningful description

Tominator87 replied:
fixed it and added description. Look if it fits your expectations

maghamed commented on app/code/Magento/Inventory/Model/SourceRepository.php:
```diff
@@ -0,0 +1,120 @@
+    public function get($sourceId)
+    {
+        /** @var SourceInterface|AbstractModel $model */
+        $model = $this->sourceFactory->create();
```
don't call variable as $model, variable names should be self-descriptive

Tominator87 replied:
fixed it

maghamed commented on app/code/Magento/Inventory/Model/SourceRepository.php:
```diff
@@ -0,0 +1,120 @@
+    {
+        /** @var \Magento\Inventory\Model\Resource\Source\Collection $collection */
+        $collection = $this->collectionFactory->create();
+        $this->collectionProcessor->process($searchCriteria, $collection);
```
have nullable $searchCriteria, you will pass null to Collection Processor.
Currently interface of Collection Processor looks like:
```php
/**
 * @api
 */
interface CollectionProcessorInterface
{
    /**
     * Apply Search Criteria to Collection
     *
     * @param SearchCriteriaInterface $searchCriteria
     * @param AbstractDb $collection
     * @throws \InvalidArgumentException
     * @return void
     */
    public function process(SearchCriteriaInterface $searchCriteria, AbstractDb $collection);
}

```
So, interface doesn't support passing null value.

Tominator87 replied:
added check and comment

maghamed commented on app/code/Magento/Inventory/Model/SourceSearchResults.php:
```diff
@@ -0,0 +1,18 @@
+
+/**
+ * Class SourceSearchResults
+ * @package Magento\Inventory\Model
```
remove package

Tominator87 replied:
fixed it

maghamed commented on app/code/Magento/Inventory/Setup/InstallSchema.php:
```diff
@@ -0,0 +1,228 @@
+    const OPTION_LENGTH = 'length';
+
+    /**
+     * @param SchemaSetupInterface $setup
```
there is no description in class

Tominator87 replied:
added description. Look if it fits your expectations

maghamed commented on app/code/Magento/Inventory/Setup/InstallSchema.php:
```diff
@@ -0,0 +1,228 @@
+            ];
+            $table->addColumn(SourceInterface::CONTACT_NAME, Table::TYPE_TEXT, 255, $options, 'Contact Name');
+
+            $options = [
```
why do we have this copy-paste for $options for all fields?

Tominator87 replied:
fixed the codestyle

maghamed commented on app/code/Magento/Inventory/Setup/InstallSchema.php:
```diff
@@ -0,0 +1,228 @@
+            $options = [
+                InstallSchema::OPTION_NULLABLE => true,
+                InstallSchema::OPTION_UNSIGNED => true,
+            ];
```
it's much easier to read such code when method chaining used , like this:
```php
       )->addColumn(
            'region',
            Table::TYPE_TEXT,
            255,
            ['nullable' => true, 'default' => null],
            'State/Province name'
        )->addColumn(
            'city',
            Table::TYPE_TEXT,
            255,
            ['nullable' => true],
            'City'
        )->addColumn(
            'street',
            Table::TYPE_TEXT,
            null,
            ['nullable' => true],
            'Street Address'
```

please use this format

Tominator87 replied:
changed it

maghamed commented on app/code/Magento/Inventory/Setup/InstallSchema.php:
```diff
@@ -0,0 +1,228 @@
+                InstallSchema::OPTION_NULLABLE => false,
+                InstallSchema::OPTION_DEFAULT => ''
+            ];
+            $table->addColumn(SourceInterface::LONGITUDE, Table::TYPE_TEXT, 255, $options, 'Longitude');
```
should be fixed on decimal

Tominator87 replied:
fixed it

maghamed commented on app/code/Magento/Inventory/Setup/InstallSchema.php:
```diff
@@ -0,0 +1,228 @@
+                InstallSchema::OPTION_NULLABLE => false,
+                InstallSchema::OPTION_DEFAULT => ''
+            ];
+            $table->addColumn(SourceInterface::LATITUDE, Table::TYPE_TEXT, 255, $options, 'Latitude');
```
decimal should be used instead

Tominator87 replied:
changed it to decimal an used 8 decimals, used this https://en.wikipedia.org/wiki/Decimal_degrees as resource. I think that should fit all thinkable usecases.

maghamed commented on app/code/Magento/Inventory/Test/Unit/Model/SourceCarrierLinkTest.php:
```diff
@@ -0,0 +1,46 @@
+    /**
+     * @var SourceCarrierLink
+     */
+    protected $sourceCarrierLink;
```
we don't encourage usage of  protected properties even in tests

Tominator87 replied:
changed it

maghamed commented on app/code/Magento/Inventory/Test/Unit/Model/SourceCarrierLinkTest.php:
```diff
@@ -0,0 +1,46 @@
+    public function testPosition()
+    {
+        $this->sourceCarrierLink->setPosition(SourceCarrierLinkTest::TEST_POSITION);
+        $this->assertEquals($this->sourceCarrierLink->getPosition(), SourceCarrierLinkTest::TEST_POSITION);
```
doesn't make sense to cover getters/setter with Unit tests.
Better to specify @ code coverage ignore

Tominator87 replied:
I have the same opinion on this topic, but is it ok to keep them? They dont't harm and add up the testcoverage, as they are already there.

ishakhsuvarov replied:
> They dont't harm and add up the testcoverage, as they are already there.

At least it may provide a negative example to other developers. Regarding the test coverage – you may mark actual methods with ignore annotation.

Commit:
magento-engcom/magento2#25 added missing unit tests in sourcetest

Commit:
Merge branch 'mni/source-management' of https://github.com/magento-engcom/magento2 into mni/source-management

maghamed commented on app/code/Magento/Inventory/Test/Unit/Model/SourceRepositoryTest.php:
```diff
@@ -0,0 +1,274 @@
+
+/**
+ * Class SourceRepositoryTest
+ * @package Magento\Inventory\Test\Unit\Model
```
remove package

Tominator87 replied:
fixed it

maghamed commented on app/code/Magento/Inventory/Model/SourceRepository.php:
```diff
@@ -0,0 +1,120 @@
+        } catch (\Exception $exception) {
+            throw new CouldNotSaveException(__($exception->getMessage()));
+        }
+        return $source;
```
repository save method should return void

to be aligned with CQRS

Tominator87 replied:
fixed it

maghamed commented on app/code/Magento/Inventory/Test/Unit/Model/SourceTest.php:
```diff
@@ -0,0 +1,139 @@
+    public function testCountryId()
+    {
+        $this->source->setCountryId(SourceTest::TEST_ID);
+        $this->assertEquals($this->source->getCountryId(), SourceTest::TEST_ID);
```
don't cover setters/getters with unit tests

Tominator87 replied:
I have the same opinion on this topic, but is it ok to keep them? They dont't harm and add up the testcoverage, as they are already there.

maghamed commented on app/code/Magento/Inventory/Test/Unit/Model/SourceRepositoryTest.php:
```diff
@@ -0,0 +1,274 @@
+
+        $result = $this->sourceRepository->save($sourceModel);
+
+        $this->assertSame($sourceModel, $result);
```
it's incorrect. because repository::save should not return value

Tominator87 replied:
fixed it

maghamed commented on app/code/Magento/InventoryApi/Api/Data/SourceInterface.php:
```diff
@@ -0,0 +1,320 @@
+
+/**
+ * @api
+ */
```
please provide a description for Interface
Each interface must have a precise and clear description

Tominator87 replied:
added description

maghamed commented on app/code/Magento/InventoryApi/Api/Data/SourceInterface.php:
```diff
@@ -0,0 +1,320 @@
+    /**
+     * Set source country id.
+     *
+     * @param string $countryId
```
int , not string

Tominator87 replied:
fixed it

maghamed commented on app/code/Magento/InventoryApi/Api/Data/SourceSearchResultsInterface.php:
```diff
@@ -0,0 +1,29 @@
+use Magento\Framework\Api\SearchResultsInterface;
+
+/**
+ * @api
```
please provide description for interface

Tominator87 replied:
added description, look if it fits your expectations

maghamed commented on app/code/Magento/InventoryApi/Api/SourceRepositoryInterface.php:
```diff
@@ -0,0 +1,45 @@
+
+/**
+ * @api
+ */
```
description needed

Tominator87 replied:
added description, look if it fits your expectations

maghamed commented on app/code/Magento/InventoryApi/Api/SourceRepositoryInterface.php:
```diff
@@ -0,0 +1,45 @@
+     * @param SourceInterface $source
+     * @return SourceInterface
+     *
+     * @throws CouldNotSaveException
```
next exception should be thrown 
    * @throws \Magento\Framework\Exception\InputException
     * @throws \Magento\Framework\Exception\StateException
     * @throws \Magento\Framework\Exception\CouldNotSaveException

maghamed commented on app/code/Magento/InventoryApi/Api/SourceRepositoryInterface.php:
```diff
@@ -0,0 +1,45 @@
+     * Save Source data.
+     *
+     * @param SourceInterface $source
+     * @return SourceInterface
```
save should return void

that's CQRS requirements

Tominator87 replied:
fixed it

nhp committed:
Remove obsolete DocBloc

maghamed commented on app/code/Magento/InventoryApi/Api/SourceRepositoryInterface.php:
```diff
@@ -0,0 +1,45 @@
+     * @throws NoSuchEntityException
+     */
+    public function get($sourceId);
+
```
please provide a description why there is no delete method in this repository.

Because community guys would be curious about this, as having delete method a general practice for magento repositories

Tominator87 replied:
added description, look if it fits your expectations

vadimjustus committed:
magento-engcom/magento2#25: WIP api-functional tests

vadimjustus committed:
Merge branch 'mni/source-management' of github.com:magento-engcom/magento2 into mni/source-management

Commit:
magento-engcom/magento2#25 made changes from codereview in pull request

Commit:
magento-engcom/magento2#25 merged changes

vadimjustus committed:
magento-engcom/magento2#25: Implement first api functional test for creating sources

vadimjustus committed:
Merge branch 'mni/source-management' of github.com:magento-engcom/magento2 into mni/source-management

vadimjustus committed:
magento-engcom/magento2#25: WIP api functional tests

ishakhsuvarov committed:
magento-engcom/magento2#25: Source API Implementation

 - Fixes according to test results

ishakhsuvarov committed:
magento-engcom/magento2#25: Source API Implementation
 - Merge branch 'develop' of github.com:magento-engcom/magento2 into mni/source-management

vadimjustus committed:
magento-engcom/magento2#25: WIP api-functional tests

vadimjustus committed:
Merge branch 'mni/source-management' of github.com:magento-engcom/magento2 into mni/source-management

vadimjustus committed:
magento-engcom/magento2#25: Finish first api-functional test

ishakhsuvarov commented on app/code/Magento/Inventory/Model/SourceRepository.php:
```diff
@@ -0,0 +1,203 @@
+    /**
+     * @param SourceInterface $source
+     */
+    private function saveSource(SourceInterface $source)
```
Is there a reason to extract this into a separate method?

ishakhsuvarov commented on app/code/Magento/Inventory/Model/SourceRepository.php:
```diff
@@ -0,0 +1,203 @@
+    /**
+     * @param SourceInterface $source
+     */
+    private function saveSourceCarrierLinks(SourceInterface $source)
```
Is there a reason to extract this into a separate method? It seems to be used just once.

ishakhsuvarov commented on dev/tests/api-functional/testsuite/Magento/InventoryApi/Api/SourceRepositoryTest.php:
```diff
@@ -0,0 +1,255 @@
+            ->setCarrierCode($carrierCode2);
+
+        /** @var  \Magento\InventoryApi\Api\Data\SourceInterface $expectedSource */
+        $expectedSource = $this->sourceFactory->create();
```
You can provide an array with data to the factory

ishakhsuvarov commented on app/code/Magento/Inventory/Test/Unit/Model/SourceRepositoryTest.php:
```diff
@@ -0,0 +1,282 @@
+
+        $this->setExpectedException(\Magento\Framework\Exception\CouldNotSaveException::class);
+
+        $this->resource->expects($this->atLeastOnce())
```
You may check if exception is logged as well

vadimjustus committed:
magento-engcom/magento2#25: wip: Implement further api-functional tests

ishakhsuvarov commented on dev/tests/api-functional/testsuite/Magento/InventoryApi/Api/SourceRepositoryTest.php:
```diff
@@ -0,0 +1,361 @@
+     */
+    private function createRandomSource($countCarrier = 2, $postcode = '54321', $isActive = true)
+    {
+        $country = $this->countryInformationAcquirer->getCountryInfo('US');
```
Integer country ID is expected by service

ishakhsuvarov committed:
magento-engcom/magento2#25: Source Management API Implementation

 - Added PHP 7 types

ishakhsuvarov commented on app/code/Magento/InventoryApi/Api/Data/SourceInterface.php:
```diff
@@ -0,0 +1,320 @@
+     * @param string $street
+     * @return SourceInterface
+     */
+    public function setStreet(string $street): SourceInterface;
```
May be better to use string[] for multi-line street fields

ishakhsuvarov committed:
magento-engcom/magento2#25: Source Management API Implementation

 - Removed PHP 7 Types to keep code PHPUnit 4 compatible
 - Fixed Unit Tests

ishakhsuvarov committed:
magento-engcom/magento2#25: Source Management API Implementation

 - Unit Tests Updated

ishakhsuvarov committed:
magento-engcom/magento2#25: Merge branch 'develop' of github.com:magento/magento2ce into mni/source-management

Commit:
Review Fixes

Commit:
magento-engcom/magento2#25 Add Bugfix for testGetSourcesList test and complete this test case

maghamed commented on app/code/Magento/Inventory/Model/Source.php:
```diff
@@ -0,0 +1,352 @@
+/**
+ * Class Source,
+ * provides implementation of the SourceInterface which adds the possibility
+ * for a Merchant to map existing physical sources to some particular sales channels
```
this interface doesn't provide an ability to

> which adds the possibility for a Merchant to map existing physical sources to some particular sales channels

for these purposes, we need to introduce a dedicated interface.

SourceToSalesChannelMapping

maghamed commented on app/code/Magento/Inventory/Model/SourceRepository.php:
```diff
@@ -0,0 +1,190 @@
+/**
+ * Class SourceRepository
+ *
+ * Provides implementation of CQRS for sourcemodel
```
the description should be fixed.

Repository doesn't provide CQRS implementation. 
It's just provide ability to retrieve entity/save and delete.

maghamed commented on app/code/Magento/Inventory/Model/SourceRepository.php:
```diff
@@ -0,0 +1,190 @@
+            foreach ($source->getCarrierLinks() as $carrierLink) {
+                $carrierLink->setData(SourceInterface::SOURCE_ID, $source->getSourceId());
+                $this->resourceSourceCarrierLink->save($carrierLink);
+            }
```
Here we have a logic problem, because we just save new sourceCarrier links, but don't remove old ones.
So, having link A and B.
We can't process change link to C
All that we can is to add C to already existing links

What we need to do , is to check that $source->getCarrierLinks is not NULL 
and if links are specified. We need to drop old links and overwrite them with new links provided in the scope of $source

```
if ($source->getCarrierLinks() !== null) {
    $connection->delete(
        $this->resourceConnection->getTableName(‘table_name_for_source_carrier_link’),
        $connection->quoteInto(‘source_id = ?‘, $source->getId())
    );
}
```

maghamed commented on app/code/Magento/Inventory/Model/SourceRepository.php:
```diff
@@ -0,0 +1,190 @@
+            /** @var SourceCarrierLinkInterface|AbstractModel $carrierLink */
+            foreach ($source->getCarrierLinks() as $carrierLink) {
+                $carrierLink->setData(SourceInterface::SOURCE_ID, $source->getSourceId());
+                $this->resourceSourceCarrierLink->save($carrierLink);
```
it's better to use Bulk insert/delete. Use Bulk inserts as methods of the resource model.
It will speed up performance 

----

I can´t finde some example Implementation in the Core please add a Example

maghamed commented on app/code/Magento/Inventory/Model/SourceRepository.php:
```diff
@@ -0,0 +1,190 @@
+    public function save(SourceInterface $source)
+    {
+        try {
+            /** @var SourceInterface|AbstractModel $source */
```
don't use AbstractModel for typing in PHPDoc, as you allowed to use just SoruceInterface methods

maghamed commented on app/code/Magento/Inventory/Model/SourceRepository.php:
```diff
@@ -0,0 +1,190 @@
+
+        $searchResults->setItems($collection->getItems());
+        $searchResults->setTotalCount($collection->getSize());
+        return $searchResults;
```
looks like current implementation of getList method would not add SourceCarrierLinks to retrieved sources

maghamed commented on app/code/Magento/Inventory/Setup/InstallSchema.php:
```diff
@@ -0,0 +1,288 @@
+     * Constant for decimal precision for latitude and longitude
+     */
+    const LATLON_PRECISION = 10;
+    const LATLON_SCALE = 6;
```
Use DECIMAL(8,6) for latitude (90 to -90 degrees) and DECIMAL(9,6) for longitude (180 to -180 degrees). 6 decimal places is fine for most applications. Both should be "signed" to allow for negative values.

maghamed commented on app/code/Magento/Inventory/Model/Source.php:
```diff
@@ -0,0 +1,352 @@
+ *
+ * @codeCoverageIgnore
+ */
+class Source extends AbstractExtensibleModel implements SourceInterface
```
Btw , need to clarify , do we need to have "Code" for Source interface.
Like code in the ERP system

maghamed commented on app/code/Magento/Inventory/Setup/InstallSchema.php:
```diff
@@ -0,0 +1,288 @@
+                    InstallSchema::OPTION_UNSIGNED => true,
+                ],
+                'Position'
+            );
```
(source_id, carrier_code) - should be compound unique index

Commit:
magento-engcom/magento2#25 Add new testcases and bugfix for SourceRepository save  method

Commit:
magento-engcom/magento2#25 Add testcase for GetSourcesList with CarrierLinks

Commit:
magento-engcom/magento2#25 Add Fixes for Review Coments

ishakhsuvarov committed:
magento-engcom/magento2#25: Source API implementation

 - Unit tests fixed

ishakhsuvarov committed:
magento-engcom/magento2#25: Merge branch 'develop' of github.com:magento/magento2ce into mni/source-management

Conflicts:
	composer.lock

ishakhsuvarov committed:
magento-engcom/magento2#25: Source API implementation

ishakhsuvarov committed:
magento-engcom/magento2#25: Source API implementation

Commit:
magento-engcom/magento2#25: Source API Implementation

-- global refactoring

ishakhsuvarov commented on app/code/Magento/InventoryApi/Api/Data/SourceInterface.php:
```diff
@@ -286,33 +286,33 @@ public function getPriority();
     /**
-     * @param \Magento\InventoryApi\Api\Data\SourceCarrierLinkInterface[] $carrierLinks
-     * @return $this
+     * @return \Magento\InventoryApi\Api\Data\SourceCarrierLinkInterface[]
```
`null` may be returned as well

Commit:
magento-engcom/magento2#25: Source API Implementation

-- global refactoring

ishakhsuvarov commented on app/code/Magento/Inventory/Model/ResourceModel/Source/Collection.php:
```diff
@@ -8,14 +8,60 @@
+    /**
+     * Collection constructor
+     *
+     * @param SourceCarrierLinkManagementInterface $sourceCarrierLinkManagement
```
Please update phpdoc with all dependencies

Commit:
magento-engcom/magento2#25: Source API Implementation

-- global refactoring
-- fix unit tests

Commit:
magento-engcom/magento2#25: Source API Implementation

-- fix static tests

Commit:
magento-engcom/magento2#25: Source API Implementation

-- fix static tests
-- fix soap tests

Commit:
magento-engcom/magento2#25: Source API Implementation

-- fix static tests

Commit:
magento-engcom/magento2#25: Source API Implementation

-- global refactoring

Commit:
magento-engcom/magento2#25: Source API Implementation

-- global refactoring

Commit:
magento-engcom/magento2#25: Source API Implementation

-- global refactoring

Commit:
magento-engcom/magento2#25: Source API Implementation

-- fix static tests

Commit:
magento-engcom/magento2#25: Source API Implementation

-- fix static tests

ishakhsuvarov committed:
magento-engcom/magento2#25: Merge branch 'develop' of github.com:magento/magento2ce into mni/source-management

Commit:
magento-engcom/magento2#25: Source API Implementation

-- fix static tests
